bindings/python/src/lib.rs

```diff
@@ -154,26 +154,68 @@ impl<'source> FromPyObject<'source> for Framework {
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum Device {
+    Cpu,
+    Cuda(usize),
+    Mps,
+}
+
+impl<'source> FromPyObject<'source> for Device {
+    fn extract(ob: &'source PyAny) -> PyResult<Self> {
+        let name: String = ob.extract()?;
+        match &name[..] {
+            "cpu" => Ok(Device::Cpu),
+            "cuda" => Ok(Device::Cuda(0)),
+            "mps" => Ok(Device::Mps),
+            name if name.starts_with("cuda:") => {
+                let tokens: Vec<_> = name.split(':').collect();
+                if tokens.len() == 2 {
+                    let device: usize = tokens[1].parse()?;
+                    Ok(Device::Cuda(device))
+                } else {
+                    Err(exceptions::PyException::new_err(format!(
+                        "device {name} is invalid"
+                    )))
+                }
+            }
+            name => Err(exceptions::PyException::new_err(format!(
```

Comment:
Can we add one last match where name can be converted to a usize? Since we can pass int/str for a device usually in PyTorch.

Reply:
Sure ! Good idea.
I already tested "mps", it works but it's slower to create tensors 1 by 1 on "mps" than creating them on CPU and then moving them to "mps".. .Don't really know why. (`torch.load(.., map_location="mps")` doesn't work too afaik)